microovn-control/src/charm.py

```diff
@@ -0,0 +1,151 @@
+                continue
+
+            # dont generate token for self or already generated
+            if self.unit.name == unit.name or unit.name in relation_data[self.unit]:
+                continue
+
+            # get hostname
+            hostname = relation_data[unit]["hostname"]
+            token = os.popen(
+                    "microovn cluster add {}".format(hostname)).read()
+            token = token.strip()
+            relation.data[self.unit][unit.name] = token
+            logger.info("added token to {}".format(unit.name))
+            newToken=True
+
+        return newToken
+
+    def __init__(self, framework: ops.Framework):
+        super().__init__(framework)
+        self._stored.set_default(in_cluster=False)
+        framework.observe(self.on.start, self._on_start)
+        framework.observe(self.on.install, self._on_install)
+        framework.observe(self.on.remove, self._on_remove)
+        if self.is_control_plane:
+            framework.observe(self.on[CONTROL_RELATION].relation_changed,
+                              self._on_peers_changed)
+        else:
+            framework.observe(self.on[WORKER_RELATION].relation_changed,
+                              self._on_cluster_changed)
+
+
+    def add_hostname(self, relation_name):
+        relation = self.model.get_relation(relation_name)
+        relation.data[self.unit]["hostname"] = os.uname()[1]
+
+    def find_token(self, relation_name):
+        if not (relation := self.model.get_relation(relation_name)):
+            return False
+
+        tokens = {
+            token
+            for unit in relation.units | {self.unit}
+            if (token := relation.data[unit].get(self.unit.name))
+        }
+        if len(tokens) != 1:
+            return False
+        (token,) = tokens
+        return token
+
+    def join_with_token(self, token):
+        self.unit.status = ops.MaintenanceStatus("Joining cluster")
+        os.system("microovn cluster join {}".format(token))
+        self._stored.in_cluster=True
+        self.unit.status = ops.ActiveStatus("Joined cluster")
+
+    def _on_install(self, event: ops.InstallEvent):
+        self.unit.status = ops.MaintenanceStatus("installing microovn snap")
+        os.system("snap install microovn --channel latest/edge")
+        self.unit.status = ops.MaintenanceStatus("waiting for microovn ready")
+        os.system("microovn waitready")
+        self.unit.status = ops.MaintenanceStatus("waiting for cluster setup")
+
+    def _on_start(self, event: ops.StartEvent):
+        """Handle start event."""
+        self.add_hostname(self.relevant_relation)
+
+        if not self.lead_control_plane:
+            return
+
+        # only bootstrap if leader
+        os.system(f"microovn cluster bootstrap")
+        self._stored.in_cluster = True
+        self.unit.status = ops.ActiveStatus("cluster bootstrapped")
+        self.update_tokens(CONTROL_RELATION)
+        self.update_tokens(WORKER_RELATION)
```

Comment:
This means that both the microovn-control and microovn-worker charms will be operating the actual microovn snap, which is a bit confusing. Let's say a human operator wants to perform a task on all MicroOVN nodes, do they have to repeat the operation across the two sets of charms? And how do we explain when/how many to deploy of each and so on?

Given the two charm strategy only exists due to limitations with scale for the peer relation, does the control charm really need to know anything about the microovn snap?

Could it be a pure charm that does not install anything, only knows how to exchange data between the control and worker charms, and then have the leader of the worker charms do the bootstrap?